source/AI.cpp

```diff
@@ -144,8 +145,36 @@ namespace {
 		{
 			// If no direct jump route, or the target system has no
 			// fuel, perform a more elaborate refueling check.
-			const DistanceMap distance(ship, to);
-			return ShouldRefuel(ship, distance);
+			vector<const System *> destinations;
+			destinations.push_back(to);
+			if(ship.GetGovernment()->IsPlayer() && player.HasTravelPlan())
+				for(const System *system : player.TravelPlan())
+					if(system != destinations.back())
+						destinations.push_back(system);
+			
+			double requiredFuel = 0.;
+			for(const System *destination : destinations)
+			{
+				const DistanceMap route(ship, from, destination);
+				
+				// Refuel if the ship cannot route further.
+				if(!route.HasRoute(destination))
```

Comment:
A "closest system" to the final system isn't calculated by DistanceMap, so the safest option is to either stay put or go to the closest system with fuel until it knows where to travel next.
Note that map distance is unacceptable for "closest system". If I landed on the Terminus wormhole and appeared in Cardea, then escorts without a quantum keystone would go to Umbral.

DistanceMap doesn't support checkpoints and doesn't support appending, which is fine for a specialized class like this one (even the search direction depends on the constructor used).
Without access to the full path, the ShouldRefuel variant with DistanceMap would only produce a meaningful result for the first section of the travel plan.
Even if it supported the other sections it would still be heavier than a copied final loop and check, which is very small.